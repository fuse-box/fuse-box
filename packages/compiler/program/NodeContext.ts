export class NodeContext {
  public namespace?: string;
  constructor() {}
}

export function createNodeContext() {
  return new NodeContext();
}
