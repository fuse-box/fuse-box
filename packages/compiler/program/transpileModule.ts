import { ASTNode } from '../interfaces/AST';
import { ICompilerOptions } from '../interfaces/ICompilerOptions';
import { FastVisit, IVisit, IVisitorMod, TopLevelVisit } from '../Visitor/Visitor';

export type ITranformerCallback = (visit: IVisit) => IVisitorMod | void | undefined;
export type ITransformer =
  | ITranformerCallback
  | {
      onTopLevelTraverse?: ITranformerCallback;
      onEachNode?: ITranformerCallback;
    };

export type ITransformerList = Array<ITransformer>;
export interface IProgramProps {
  ast: ASTNode;
  globalContext?: any;
  namespace?: string;
  transformers: ITransformerList;
  compilerOptions?: ICompilerOptions;
}

export function transpileModule(props: IProgramProps) {
  const eachNodeTransformers: Array<ITranformerCallback> = [];
  const onTopLevelTransformers: Array<ITranformerCallback> = [];
  //console.log(JSON.stringify(props.ast, null, 2));
  for (const t of props.transformers) {
    if (typeof t === 'object') {
      if (t.onEachNode) eachNodeTransformers.push(t.onEachNode);
      if (t.onTopLevelTraverse) onTopLevelTransformers.push(t.onTopLevelTraverse);
    } else {
      eachNodeTransformers.push(t);
    }
  }

  props.globalContext.programProps = props;
  if (onTopLevelTransformers.length) {
    TopLevelVisit({
      ast: props.ast,
      globalContext: props.globalContext,
      fn: visit => {
        for (const onTopLevelCallback of onTopLevelTransformers) {
          let response = onTopLevelCallback(visit);
          if (response) return response;
        }
      },
    });
  }
  FastVisit({
    ast: props.ast,
    globalContext: props.globalContext,
    fn: visit => {
      for (const eachNodeCallback of eachNodeTransformers) {
        let response = eachNodeCallback(visit);
        if (response) return response;
      }
    },
  });
  if (props.globalContext.completeCallbacks.length) {
    for (const cb of props.globalContext.completeCallbacks) {
      cb();
    }
  }
}
