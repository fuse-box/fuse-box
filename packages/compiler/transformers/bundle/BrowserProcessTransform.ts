import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { isPropertyOrPropertyAccess } from '../../Visitor/helpers';

interface BrowserProcessTransformProps {
  env?: { [key: string]: string };
}
export type IBrowserProcessTransform = BrowserProcessTransformProps & ITransformerSharedOptions;
export function BrowserProcessTransform(options: IBrowserProcessTransform) {
  options = options || {};
  return (visit: IVisit): IVisitorMod => {
    const { node, parent } = visit;

    const accessList = isPropertyOrPropertyAccess(node, parent, 'process');
    if (accessList) {
      //console.log(accessList, node);
    }
    return;
  };
}
