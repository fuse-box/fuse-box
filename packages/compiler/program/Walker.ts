import { NodeContext } from "./NodeContext";

export interface IVisitProps {
  parent?: any;
  prop?: string;
  idx?: number;
  context?: IASTContext;
}
export interface IASTContext {
  nodeContext?: NodeContext;
  locals?: Array<string>;
  exports?: Array<string>;
}

export interface IASTVisitor {
  node: any;
  props: IVisitProps;
  context?: IASTContext;
}
interface IASTWalkProps {
  withScope?: boolean;
  visit?: (props: IASTVisitor) => any;
}
export function fastWalk(ast: any, walker: IASTWalkProps) {
  function visit(node, props: IVisitProps, context) {
    if (walker.withScope) {
      if (node.context) {
        context = node.context;
        if (!node.context.locals) node.context.locals = [];
      }

      if (node.type === "VariableDeclaration") {
        if (node.declarations) {
          for (const decl of node.declarations) {
            if (
              decl.type === "VariableDeclarator" &&
              decl.id &&
              decl.id.type === "Identifier"
            ) {
              if (context === undefined) context = { locals: [] };
              context.locals.push(decl.id.name);
              // we need to check for the next item on the list (if we are in an array)
              if (props.idx && props.prop) {
                if (props.parent[props.prop][props.idx + 1]) {
                  props.parent[props.prop][props.idx + 1].context = context;
                }
              }
            }
          }
        }
      }

      if (node.type === "ExpressionStatement") {
        if (node.expression && node.expression.arguments) {
          node.expression.arguments.context = context;
        }
      }

      if (
        node.type === "FunctionDeclaration" ||
        node.type === "FunctionExpression"
      ) {
        if (node.id && node.id.type === "Identifier") {
          if (context === undefined) context = { locals: [] };
          context.locals.push(node.id.name);
          if (props.idx && props.prop) {
            if (props.parent[props.prop][props.idx + 1]) {
              props.parent[props.prop][props.idx + 1].context = context;
            }
          }
        }
        if (node.params) {
          for (const item of node.params) {
            if (item.type === "Identifier") {
              if (node.body.context === undefined) {
                // create context
                node.body.context = {
                  exports:
                    context && context.exports
                      ? context.exports.concat([])
                      : [],
                  // here we need to make a copy of the locals
                  locals: context ? context.locals.concat([]) : []
                };
              }
              node.body.context.locals.push(item.name);
            }
          }
        }
      }
    }

    const newNode = walker.visit({ node, props, context });
    if (newNode) {
      if (props.prop && props.parent) {
        if (props.idx) {
          props.parent[props.prop][props.idx] = newNode;
        } else {
          props.parent[props.prop] = newNode;
        }
      }
      node = newNode;
    }
    for (const prop in node) {
      if (prop[0] === "$") {
        continue;
      }
      const child = node[prop];
      if (child instanceof Array) {
        for (let i = 0; i < child.length; i++) {
          if (child && child[i] && child[i].type) {
            visit(child[i], { parent: node, prop, idx: i }, context);
          }
        }
      } else {
        if (child && child.type) visit(child, { parent: node, prop }, context);
      }
    }
  }

  visit(ast, {}, undefined);
}

export interface IVisit {
  node: any;
  parent?: any;
  prop?: string;
  idx?: number;
  context?: IASTContext;
}
export function FastVisit(ast, fn: (visit: IVisit) => void) {
  function _visit(
    node,
    props: { parent?; propName?; id? },
    context: IASTContext
  ) {}

  _visit(ast, {}, undefined);
}
