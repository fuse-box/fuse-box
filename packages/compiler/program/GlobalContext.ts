import { ASTNode } from '../interfaces/AST';
import { IProgramProps } from './transpileModule';

export interface GlobalContext {
  getNextIndex: () => number;
  exportAfterDeclaration?: {
    [key: string]: {
      target?: string;
    };
  };
  programProps?: IProgramProps;
  identifierReplacement: {
    [key: string]: {
      first?: string;
      second?: string;
      insertAfter?: ASTNode;
      inUse?: boolean;
    };
  };
  namespace?: string;
  completeCallbacks?: Array<() => void>;
}

export function createGlobalContext(userContext?: { [key: string]: any }): GlobalContext {
  let index = 1;
  let essentialContext = {
    completeCallbacks: [],
    getNextIndex: () => index++,
    identifierReplacement: {},
    namespace: 'exports',
  };
  if (userContext) {
    for (const key in userContext) {
      essentialContext[key] = userContext[key];
    }
  }

  return essentialContext;
}
