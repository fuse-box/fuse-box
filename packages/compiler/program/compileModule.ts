import * as buntis from "buntis";
import { generate } from "../generator/generator";
import { ASTNode } from "../interfaces/AST";
import { ICompilerOptions } from "../interfaces/ICompilerOptions";
import { ClassConstructorPropertyTransformer } from "../transformers/ClassConstructorPropertyTransformer";
import { ExportTransformer } from "../transformers/ExportTransformer";
import { GlobalContextTransformer } from "../transformers/GlobalContextTransformer";
import { ImportTransformer } from "../transformers/ImportTransformer";
import { InterfaceRemoverTransformer } from "../transformers/InterfaceRemoverTransformer";
import { JSXTransformer } from "../transformers/JSXTransformer";
import { NamespaceTransformer } from "../transformers/NameSpaceTransformer";
import { IVisit, IVisitorMod } from "../Visitor/Visitor";
import { createGlobalContext } from "./GlobalContext";
import { ITransformerList, transpileModule } from "./transpileModule";

export interface ICompileModuleProps {
  code: string;
  globalContext?: any;
  transformers?: Array<(globalContext) => (visit: IVisit) => IVisitorMod>;
  compilerOptions?: ICompilerOptions;
}

export function compileModule(props: ICompileModuleProps) {
  const ast = buntis.parseTSModule(props.code, {
    directives: true,
    jsx: true,
    next: true,
    loc: true,
    ts: true
  });

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    ClassConstructorPropertyTransformer(),
    JSXTransformer(),
    NamespaceTransformer(),
    InterfaceRemoverTransformer(),
    ImportTransformer(),
    ExportTransformer()
  ];
  transpileModule({
    ast: ast as ASTNode,
    compilerOptions: props.compilerOptions,
    globalContext: createGlobalContext(props.globalContext),
    transformers: defaultTransformers
  });
  //console.log(JSON.stringify(ast, null, 2));
  const res = generate(ast, {});

  return { code: res };
}
