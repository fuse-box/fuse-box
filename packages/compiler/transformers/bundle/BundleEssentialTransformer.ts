import * as path from 'path';
import { ASTNode } from '../../interfaces/AST';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { GlobalContext } from '../../program/GlobalContext';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';

export interface IBundleEssentialProps {
  target?: 'browser' | 'server' | 'electron';
  env?: { [key: string]: any };
  moduleFileName?: string;
}

export type IBundleEssntialTransformerOptions = IBundleEssentialProps & ITransformerSharedOptions;

export const PolyfillEssentialConfig = {
  stream: 'stream',
  buffer: 'buffer',
  Buffer: 'buffer',
  http: 'http',
  https: 'https',
};

export function BundleEssentialTransformer(options: IBundleEssntialTransformerOptions) {
  const RequireStatementsInserted = {};
  const VariablesInserted = {};

  return (visit: IVisit): IVisitorMod => {
    const { node } = visit;

    if (visit.isLocalIdentifier) {
      const globalContext = visit.globalContext as GlobalContext;
      const locals = visit.scope && visit.scope.locals ? visit.scope.locals : {};
      const name = node.name;

      if (locals[name] || globalContext.hoisted[name]) {
        return;
      }

      switch (name) {
        case '__dirname':
          return { replaceWith: { type: 'Literal', value: options.moduleFileName } };
        case '__filename':
          return { replaceWith: { type: 'Literal', value: path.dirname(options.moduleFileName) } };
      }

      if (options.target !== 'browser') return;

      /**
       * *********************************************************************************
       * Polyfills stream buffer e.t.c
       */
      if (PolyfillEssentialConfig[name]) {
        if (VariablesInserted[name]) return;
        const statements: Array<ASTNode> = [];
        const moduleName = PolyfillEssentialConfig[name];
        if (!RequireStatementsInserted[moduleName]) {
          RequireStatementsInserted[moduleName] = 1;
          const statement = createRequireStatement(moduleName, name);
          statements.push(statement.statement);
        }
        if (name !== moduleName) {
          statements.push({
            type: 'VariableDeclaration',
            kind: 'var',
            declarations: [
              {
                type: 'VariableDeclarator',
                init: {
                  type: 'Identifier',
                  name: moduleName,
                },
                id: {
                  type: 'Identifier',
                  name: name,
                },
              },
            ],
          });
        }
        if (statements.length) {
          return { prependToBody: statements };
        }
      }

      /**
       * Process transformation **********************************************************
       */
      if (name === 'process') {
        const parent = visit.parent;
      }
    }
    return;
  };
}
