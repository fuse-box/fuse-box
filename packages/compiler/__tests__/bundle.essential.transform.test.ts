import { compileModule } from '../program/compileModule';
import { PolyfillEssentialConfig } from '../transformers/bundle/BundleEssentialTransformer';
describe('Bundle essential transform test', () => {
  describe('Common transform ', () => {
    it('should replace __dirname', () => {
      const result = compileModule({
        code: `
           console.log(__dirname)
      `,
        bundleProps: {
          moduleFileName: '/some-dir/file.ts',
        },
      });
      expect(result.code).toMatchInlineSnapshot(`
                "console.log(\\"/some-dir/file.ts\\");
                "
            `);
    });

    it('should replace __filename', () => {
      const result = compileModule({
        code: `
           console.log(__filename)
      `,
        bundleProps: {
          moduleFileName: '/some-dir/file.ts',
        },
      });
      expect(result.code).toMatchInlineSnapshot(`
        "console.log(\\"/some-dir\\");
        "
      `);
    });
  });

  describe('polyfills', () => {
    describe('Single reference', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName}`, () => {
          const result = compileModule({
            code: `console.log(${moduleName})`,
          });
          expect(result.code).toMatchSnapshot();
        });
      }
    });

    describe('Member reference reference', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName} with method reference`, () => {
          const result = compileModule({
            code: `console.log(${moduleName}.method)`,
          });

          expect(result.code).toMatchSnapshot();
        });
      }
    });
    describe('Member reference with a call', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName} with method reference with a call`, () => {
          const result = compileModule({
            code: `console.log(${moduleName}.method() )`,
          });

          expect(result.code).toMatchSnapshot();
        });
      }
    });

    describe('Should not add anything', () => {
      it('should not add Buffer because its been hoisted', () => {
        const result = compileModule({
          code: `
            exports.Buffer = Buffer;
            function Buffer(){}
          `,
        });

        expect(result.code).toMatchSnapshot();
      });

      it('should not add Buffer because its has been defined', () => {
        const result = compileModule({
          code: `
            function some(){
              const buffer = {};
              console.log(buffer)
            }
          `,
        });
        expect(result.code).toMatchSnapshot();
      });

      it('should not add Buffer because it has NOT been defined', () => {
        const result = compileModule({
          code: `
            function some(){
              console.log(buffer)
            }
          `,
        });

        expect(result.code).toMatchSnapshot();
      });
    });
  });
});
