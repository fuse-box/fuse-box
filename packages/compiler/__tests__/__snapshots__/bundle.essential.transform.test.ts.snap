// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle essential transform test polyfills Member reference reference shuold insert buffer with method reference 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method);
"
`;

exports[`Bundle essential transform test polyfills Member reference reference shuold insert buffer with method reference 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method);
"
`;

exports[`Bundle essential transform test polyfills Member reference reference shuold insert http with method reference 1`] = `
"var http = require(\\"http\\");
console.log(http.method);
"
`;

exports[`Bundle essential transform test polyfills Member reference reference shuold insert https with method reference 1`] = `
"var https = require(\\"https\\");
console.log(https.method);
"
`;

exports[`Bundle essential transform test polyfills Member reference reference shuold insert stream with method reference 1`] = `
"var stream = require(\\"stream\\");
console.log(stream.method);
"
`;

exports[`Bundle essential transform test polyfills Member reference with a call shuold insert buffer with method reference with a call 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method());
"
`;

exports[`Bundle essential transform test polyfills Member reference with a call shuold insert buffer with method reference with a call 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method());
"
`;

exports[`Bundle essential transform test polyfills Member reference with a call shuold insert http with method reference with a call 1`] = `
"var http = require(\\"http\\");
console.log(http.method());
"
`;

exports[`Bundle essential transform test polyfills Member reference with a call shuold insert https with method reference with a call 1`] = `
"var https = require(\\"https\\");
console.log(https.method());
"
`;

exports[`Bundle essential transform test polyfills Member reference with a call shuold insert stream with method reference with a call 1`] = `
"var stream = require(\\"stream\\");
console.log(stream.method());
"
`;

exports[`Bundle essential transform test polyfills Should not add anything should not add Buffer because it has NOT been defined 1`] = `
"var buffer = require(\\"buffer\\");
function some() {
  console.log(buffer);
}
"
`;

exports[`Bundle essential transform test polyfills Should not add anything should not add Buffer because its been hoisted 1`] = `
"exports.Buffer = Buffer;
function Buffer() {}
"
`;

exports[`Bundle essential transform test polyfills Should not add anything should not add Buffer because its has been defined 1`] = `
"function some() {
  const buffer = {};
  console.log(buffer);
}
"
`;

exports[`Bundle essential transform test polyfills Single reference shuold insert buffer 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer);
"
`;

exports[`Bundle essential transform test polyfills Single reference shuold insert buffer 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer);
"
`;

exports[`Bundle essential transform test polyfills Single reference shuold insert http 1`] = `
"var http = require(\\"http\\");
console.log(http);
"
`;

exports[`Bundle essential transform test polyfills Single reference shuold insert https 1`] = `
"var https = require(\\"https\\");
console.log(https);
"
`;

exports[`Bundle essential transform test polyfills Single reference shuold insert stream 1`] = `
"var stream = require(\\"stream\\");
console.log(stream);
"
`;
